### ISSUE TITLE
Keras .fit() yields incorrect results when using a custom loss function

### ISSUE THREAD

**Issue by netw0rkf10w** [tensorflow/tensorflow#33988]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- TensorFlow version (use command below): 2.0.0

**Describe the current behavior**
When using a custom loss function, Keras' `.fit()` produces incorrect results.

**Code to reproduce the issue**
See this [Colab Notebook](https://colab.research.google.com/drive/1Q5hVfxVKdeqtIPWrYuvsPQ-IE0dUOeWO).

**Comment by oanush** [comment]
Issue replicating for TF-2.0.

**Comment by jvishnuvardhan** [comment]
@netw0rkf10w Can you please provide simple standalone code to reproduce the issue. It is really faster  to resolve with simple standalone codes. Thanks!

**Comment by netw0rkf10w** [comment]
@jvishnuvardhan To show that the results of `.fit()` are incorrect, I had to add GradientTape for comparison. Here's [a notebook without custom training loop](https://colab.research.google.com/drive/1eNAFiOY8fDWMF4ydx7-RIoc8Fi7gRia_), but then you'll have to figure out why the results are not correct.
Thanks.

**Comment by netw0rkf10w** [comment]
FYI when using a different custom loss function, derived directly from `tf.keras.losses.Loss`, then Keras `.fit()`'s results are consistent with GradientTape's:

```
class SparseCategoricalCrossentropyIgnoreLabel(tf.keras.losses.Loss):
    """Computes the crossentropy loss between the labels and predictions,
        with ignored label.
        Derived directly from tf.keras.losses.Loss
    """
    def __init__(self, ignore_label=None, from_logits=False,
                 reduction=tf.keras.losses.Reduction.AUTO,
                 name='sparse_categorical_crossentropy_ignore_label'):
        super().__init__(reduction=reduction, name=name)
        self.ignore_label = ignore_label
        self.from_logits = from_logits

    def call(self, y_true, y_pred):
        """
        Keep only batch dimension for the final loss
        """
        y_true = tf.reshape(y_true, (tf.shape(y_true)[0], -1, tf.shape(y_true)[-1]))
        y_pred = tf.reshape(y_pred, (tf.shape(y_pred)[0], -1, tf.shape(y_pred)[-1]))
        loss_tensor = tf.keras.backend.sparse_categorical_crossentropy(
                y_true, y_pred,
                from_logits=self.from_logits)
        mask = tf.ones_like(loss_tensor)
        if self.ignore_label is not None:
            mask = (y_true != self.ignore_label)
            mask = tf.squeeze(mask, [-1])
            loss_tensor = tf.where(mask, loss_tensor, 0)
        
        num_elements = tf.reduce_sum(tf.cast(mask, tf.float32), axis=1)
        loss_tensor = tf.reduce_sum(loss_tensor, axis=1)/num_elements
        loss_tensor = tf.where(num_elements > 0, loss_tensor, 0)
        return loss_tensor
```

The loss function defined in the above Colab notebook (i.e. the one that caused the issue) is derived from `tf.keras.losses.SparseCategoricalCrossentropy`:

```
class SparseCategoricalCrossentropyIgnoreLabel(tf.keras.losses.SparseCategoricalCrossentropy):
    """Computes the crossentropy loss between the labels and predictions,
        with ignored label.
        Derived from tf.keras.losses.SparseCategoricalCrossentropy
    """
    def __init__(self, ignore_label=None, from_logits=False,
                 reduction=tf.keras.losses.Reduction.AUTO,
                 name='sparse_categorical_crossentropy_ignore_label'):
        super().__init__(from_logits=from_logits, reduction=reduction, name=name)
        self.ignore_label = ignore_label
        
    def __call__(self, y_true, y_pred, sample_weight=None):
        if self.ignore_label is not None:
            if sample_weight is not None:
                sample_weight = tf.where(y_true == self.ignore_label, 0, sample_weight)
            else:
                sample_weight = y_true != self.ignore_label
        return super().__call__(y_true, y_pred, sample_weight)
```

**Comment by jvishnuvardhan** [comment]
@netw0rkf10w With `tf-nightly` I am seeing loss values are similar. Please take a look at the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/0fc2c5a7aeaff8e0ee0bf78e1934584c/tf_custom_loss.ipynb). Please let us know what you think.

Here is the loss from `model.fit`

```
Epoch 1/5
19/19 [==============================] - 16s 854ms/step - loss: 0.9181 - sparse_categorical_accuracy: 0.5864 - val_loss: 0.8268 - val_sparse_categorical_accuracy: 0.6771
Epoch 2/5
19/19 [==============================] - 15s 794ms/step - loss: 0.5648 - sparse_categorical_accuracy: 0.7903 - val_loss: 0.6254 - val_sparse_categorical_accuracy: 0.7927
Epoch 3/5
19/19 [==============================] - 15s 778ms/step - loss: 0.4492 - sparse_categorical_accuracy: 0.8387 - val_loss: 0.5172 - val_sparse_categorical_accuracy: 0.8288
Epoch 4/5
19/19 [==============================] - 15s 774ms/step - loss: 0.3835 - sparse_categorical_accuracy: 0.8566 - val_loss: 0.4660 - val_sparse_categorical_accuracy: 0.8445
Epoch 5/5
19/19 [==============================] - 15s 775ms/step - loss: 0.3476 - sparse_categorical_accuracy: 0.8679 - val_loss: 0.4404 - val_sparse_categorical_accuracy: 0.8488
```

Here is the loss from custom training

```
Epoch 1/5
19/19 [==============================] - 18s 970ms/step - loss: 0.4031 - sparse_categorical_accuracy: 0.3776 - val_loss: 0.2922 - val_sparse_categorical_accuracy: 0.3049
Epoch 2/5
19/19 [==============================] - 15s 784ms/step - loss: 0.2247 - sparse_categorical_accuracy: 0.3220 - val_loss: 0.2261 - val_sparse_categorical_accuracy: 0.3394
Epoch 3/5
19/19 [==============================] - 15s 777ms/step - loss: 0.1888 - sparse_categorical_accuracy: 0.3448 - val_loss: 0.1981 - val_sparse_categorical_accuracy: 0.3546
Epoch 4/5
19/19 [==============================] - 15s 778ms/step - loss: 0.1750 - sparse_categorical_accuracy: 0.3530 - val_loss: 0.1959 - val_sparse_categorical_accuracy: 0.3566
Epoch 5/5
19/19 [==============================] - 15s 778ms/step - loss: 0.1657 - sparse_categorical_accuracy: 0.3522 - val_loss: 0.1928 - val_sparse_categorical_accuracy: 0.3584
```

**Comment by netw0rkf10w** [comment]
Sorry but the loss values you showed are **clearly** not similar. Just compare the two sections **Using built-in loss function** and **Using custom loss function** in the [gist](https://colab.research.google.com/drive/1Q5hVfxVKdeqtIPWrYuvsPQ-IE0dUOeWO) to see the difference.

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF v2.5 ,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/9c0e17d896996cdc76614dfb72465fd0/untitled26.ipynb)..Thanks !

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
